text/0000-pbft-regular-view-change.md

```diff
@@ -0,0 +1,185 @@
+protocol perfectly but manipulate the ordering to its benefit by:
+
+1. Manipulating the order of a given set of transactions
+2. Generating and inserting new transactions into the ordering
+3. Withholding submitted transactions from the ordering
+
+One thing that makes this problem hard is that, without application-specific
+knowledge from the transactions, it is difficult to tell if an ordering
+actually benefits the leader or if "random noise" caused the ordering to be
+substantially different than some expected or fair ordering.
+
+The current implementation of PBFT does not prevent a leader from unfairly
+ordering transactions. This is a problem when a leader node has an incentive to
+do so. This RFC aims to mitigate this problem in the interest of making PBFT
+more resilient to bad actors.
+
+This RFC also proposes a mitigating solution to the "silent leader" problem.
+Sawtooth PBFT checks for leader liveness by starting a timer when a leader
+proposes a new block and endorses it with a pre-prepare message. If the timer
+expires before the new block is committed, the leader is suspected of being
+faulty and a view change is initiated. However, if a leader never proposes a
+block or doesn't endorse a block with a pre-prepare message, no timer is
+started. This means that a faulty leader can "remain silent" and stall the
+network without the other nodes being able to determine whether the silence is
+because no new batches are arriving, or the leader is intentionally ignoring
+them.
+
+                PrePreparing  <----------- Finished
+    (start timer) |                           ^
+                  v                           | (stop timer)
+                Preparing -------------> Committing
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In order to mitigate the effects of the "unfair ordering" and "silent leader"
+problems, Sawtooth PBFT will perform view changes in two new cases.
+
+In the first case, regular view changes are "forced" based on the number of
+committed blocks. This mitigates the "unfair ordering" problem by giving every
+node a chance to be unfair for a period of blocks. This is inspired by how the
+Tendermint consensus algorithm handles the problem and how lottery-style
+algorithms handle the problem, where a new leader can be elected for every
+block.
+
+In the second case, nodes propose a standard view change when the network is
+idle for too long (the primary does not produce a block and a pre-prepare
+message). This mitigates the "silent leader" problem by limiting the amount of
+time a faulty leader can stall the network.
+
+The setting `sawtooth.consensus.pbft.forced_view_change_period` determines how
+often, measured in blocks, a view change should be forced. After transitioning
+from the Finished to PrePreparing phase, a node will check whether a view change
+should be forced and, if so, immediately change views. Specifically, the node
```

Comment:
My read of the Castro paper is that the backup proposing a view change will be the leader (primary) of the new view. Which node would be elected in this case? Would it be deterministic, and if so does that predictability present a new and meaningful risk?

Reply:
That is incorrect; the new primary is deterministically calculated based on the ordering of the nodes (in the `sawtooth.consensus.pbft.peers` setting) and the view number. If you have _n_ nodes that are numbered _0 .. n - 1_, node _p_ will be the primary as determined by the formula: _p = v mod n_, where _v_ is the view number.

Reply:
So basically, when you force a view change (increment the view by 1), it will just be the next node in the on-chain list that becomes the primary.

Reply:
Yes, I was incorrect. The sequencing you describe is consistent with the paper.